ast/compile_test.go

```diff
@@ -5135,6 +5134,124 @@ func TestCheckUnusedAssignedAndArgVars(t *testing.T) {
 			}`,
 			expectedErrors: Errors{},
 		},
+		{
+			note: "argvar not used in body but in head value comprehension",
+			module: `package test
+			a := {"foo": 1}
+			func(x) := { x: v | v := a[x] } {
+				input.test == "foo"
+			}`,
+			expectedErrors: Errors{},
+		},
+		{
+			note: "argvar not used in body and shadowed in head value comprehension",
+			module: `package test
+			a := {"foo": 1}
+			func(x) := { x: v | x := "foo"; v := a[x] } {
+				input.test == "foo"
+			}`,
+			expectedErrors: Errors{
+				&Error{
+					Code:     CompileErr,
+					Location: NewLocation([]byte("func(x) := { x: v | x := \"foo\"; v := a[x] }"), "", 3, 4),
+					Message:  "unused argument x",
+				},
+			},
+		},
+		{
+			note: "argvar used in primary body but not in else body",
+			module: `package test
+			func(x) {
+				input.test == x
+			} else := false {
+				input.test == "foo"
+			}`,
+			expectedErrors: Errors{},
+		},
+		{
+			note: "argvar not used in primary body but in else body",
+			module: `package test
+			func(x) {
+				input.test == "foo"
+			} else := false {
+				input.test == x
+			}`,
+			expectedErrors: Errors{},
+		},
+		{
+			note: "argvar used in primary body but not in implicit else body",
+			module: `package test
+			func(x) {
+				input.test == x
+			} else := false`,
+			expectedErrors: Errors{},
+		},
+		{
+			note: "argvars usage spread over multiple bodies",
+			module: `package test
+			func(x, y, z) {
+				input.test == x
+			} else {
+				input.test == y
+			} else {
+				input.test == z
+			}`,
+			expectedErrors: Errors{},
+		},
+		{
+			note: "argvars usage spread over multiple bodies, missing in first",
+			module: `package test
+			func(x, y, z) {
+				input.test == "foo"
+			} else {
+				input.test == y
+			} else {
+				input.test == z
+			}`,
+			expectedErrors: Errors{
+				&Error{
+					Code:     CompileErr,
+					Location: NewLocation([]byte("func(x, y, z)"), "", 2, 4),
+					Message:  "unused argument x",
+				},
+			},
+		},
+		{
+			note: "argvars usage spread over multiple bodies, missing in second",
+			module: `package test
+			func(x, y, z) {
+				input.test == x
+			} else {
+				input.test == "bar"
+			} else {
+				input.test == z
+			}`,
+			expectedErrors: Errors{
+				&Error{
+					Code:     CompileErr,
+					Location: NewLocation([]byte("func(x, y, z)"), "", 2, 4),
+					Message:  "unused argument y",
+				},
+			},
+		},
+		{
+			note: "argvars usage spread over multiple bodies, missing in third",
+			module: `package test
+			func(x, y, z) {
+				input.test == x
+			} else {
+				input.test == y
+			} else {
+				input.test == "baz"
+			}`,
+			expectedErrors: Errors{
+				&Error{
+					Code:     CompileErr,
+					Location: NewLocation([]byte("func(x, y, z)"), "", 2, 4),
+					Message:  "unused argument z",
+				},
+			},
+		},
```

Comment:
🤔 Do we support ref-head functions? It would perhaps be good to have an extra test case, like
```rego
package test
a.b.c.f(x) = y { ... }
```

Reply:
I'll add a test for this 👍 .

Reply:
Also added tests for else-branching funcs with wildcards, to be safe.